File: leetcode/2147_Number_of_Ways_to_Divide_a_Long_Corridor.py
from typing import Optional
from unittest import TestCase

class Solution:
    def numberOfWays(self, corridor: str) -> int:
        first_seat, last_seat, seat_cnt = None, None, 0 
        for (i, val) in enumerate(corridor):
            if val == 'S':
                last_seat = i
                seat_cnt += 1
                if first_seat is None:
                    first_seat = i
        if seat_cnt > 0 and seat_cnt % 2 == 0:
            cnt, seat_cnt, cont_plant_cnt = 1, 0, 0
            for i in range(first_seat, last_seat):
                if corridor[i] == 'S':
                    seat_cnt += 1
                    if cont_plant_cnt > 0:
                        cnt = cnt * (cont_plant_cnt + 1) % 1000000007
                        cont_plant_cnt = 0
                elif seat_cnt % 2 == 0:
                    cont_plant_cnt += 1
            return cnt
        return 0
        
class NumberOfWaysTest(TestCase):
    def test_1(self):
        self.assertEqual(Solution().numberOfWays("SSPPSPS"), 3)
        
    def test_2(self):
        self.assertEqual(Solution().numberOfWays("PPSPSP"), 1)

    def test_3(self):
        self.assertEqual(Solution().numberOfWays("S"), 0)

    def test_4(self):
        self.assertEqual(Solution().numberOfWays("SPSPPSSPSSSS"), 6)

File: leetcode/2147_Number_of_Ways_to_Divide_a_Long_Corridor.rs
struct Solution {}

#[derive(PartialEq)]
enum Item {
    Seat,
    Plant,
}

impl Solution {
    pub fn number_of_ways(corridor: String) -> i32 {
        let corridor = corridor.chars().map(|char| if char == 'S' { Item::Seat } else { Item::Plant } ).collect::<Vec<_>>();
        let mut first_seat = None;
        let mut last_seat = 0;
        let mut seat_cnt = 0;
        for (i, item) in corridor.iter().enumerate() {
            if *item == Item::Seat {
                seat_cnt += 1;
                last_seat = i;
                if first_seat.is_none() {
                    first_seat = Some(i);
                }
            }
        }
        if let Some(first_seat) = first_seat {
            if seat_cnt % 2 == 0 {
                let mut seat_cnt: u64 = 0;
                let mut plant_cnt: u64 = 0;
                let mut cnt: u64 = 1;
                for i in first_seat..last_seat {
                    if corridor[i] == Item::Seat {
                        seat_cnt += 1;
                        if plant_cnt > 0 {
                            cnt = cnt * (plant_cnt + 1) % 1000000007;
                            plant_cnt = 0;
                        }
                    } else {
                        if seat_cnt % 2 == 0 {
                            plant_cnt += 1;
                        }
                    }
                }
                return cnt as i32;
            }
        }
        0
    }
}

#[test]
fn test_1() {
    assert_eq!(3, Solution::number_of_ways("SSPPSPS".to_string()))
}

#[test]
fn test_2() {
    assert_eq!(1, Solution::number_of_ways("PPSPSP".to_string()))
}

#[test]
fn test_3() {
    assert_eq!(0, Solution::number_of_ways("S".to_string()))
}